js/modules/k6/grpc/client.go

```diff
@@ -0,0 +1,479 @@
+)
+
+// Client reprecents a gRPC client that can be used to make RPC requests
+type Client struct {
+	mds  map[string]*desc.MethodDescriptor
+	tags map[string]string
+	conn *grpc.ClientConn
+}
+
+// NewClient creates a new gPRC client to make invoke RPC methods.
+func (*GRPC) NewClient(ctxPtr *context.Context) interface{} {
+	rt := common.GetRuntime(*ctxPtr)
+
+	return common.Bind(rt, &Client{}, ctxPtr)
+}
+
+// MethodInfo holds ifromation on any parsed method descriptors that can be used by the goja VM
+type MethodInfo struct {
+	FullMethod     string
+	IsClientStream bool
+	IsServerStream bool
+}
+
+// Response is a gRPC response that can be used by the goja VM
+type Response struct {
+	Status   codes.Code
+	Message  goja.Value
+	Headers  map[string][]string
+	Trailers map[string][]string
+	Error    goja.Value
+}
+
+// Load will parse the given proto files and make the file descriptors available to request.
+func (c *Client) Load(ctxPtr *context.Context, importPaths []string, filenames ...string) ([]MethodInfo, error) {
+	if lib.GetState(*ctxPtr) != nil {
+		return nil, errors.New("load must be called in the init context")
+	}
+
+	parser := protoparse.Parser{
+		ImportPaths:      importPaths,
+		InferImportPaths: len(importPaths) == 0,
+	}
+
+	fds, err := parser.ParseFiles(filenames...)
+	if err != nil {
+		return nil, err
+	}
+
+	var rtn []MethodInfo
+	c.mds = make(map[string]*desc.MethodDescriptor)
+	for _, fd := range fds {
+		for _, sd := range fd.GetServices() {
+			for _, md := range sd.GetMethods() {
+				var s strings.Builder
+				s.WriteString(sd.GetFullyQualifiedName())
+				s.WriteRune('/')
+				s.WriteString(md.GetName())
+				name := s.String()
+				c.mds[name] = md
+				rtn = append(rtn, MethodInfo{
+					FullMethod:     name,
+					IsClientStream: md.IsClientStreaming(),
+					IsServerStream: md.IsServerStreaming(),
+				})
+			}
+		}
+	}
+
+	return rtn, nil
+}
+
+type transportCreds struct {
+	credentials.TransportCredentials
+	errc chan<- error
+}
+
+func (t transportCreds) ClientHandshake(ctx context.Context,
+	addr string, in net.Conn) (net.Conn, credentials.AuthInfo, error) {
+	out, auth, err := t.TransportCredentials.ClientHandshake(ctx, addr, in)
+	if err != nil {
+		t.errc <- err
+	}
+
+	return out, auth, err
+}
+
+// Connect is a block dial to the gRPC server at the given address (host:port)
+func (c *Client) Connect(ctxPtr *context.Context, addr string, params map[string]interface{}) (bool, error) {
+	state := lib.GetState(*ctxPtr)
+	if state == nil {
+		return false, errConnectInInitContext
+	}
+
+	isPlaintext, timeout := false, 60*time.Second
+
+	for k, v := range params {
+		switch k {
+		case "plaintext":
+			isPlaintext, _ = v.(bool)
+		case "timeout":
+			if t, ok := v.(float64); ok && t > 0.0 {
+				timeout = time.Duration(t) * time.Millisecond
+			}
+		}
+	}
+
+	errc := make(chan error)
+	go func() {
```

Comment:
This is needed to report on TLS errors: Even with `FailOnNonTempDialError`, if there is a TLS error this will timeout rather than report the error, so we can't rely on `WithBlock`. By running in a goroutine we can then wait on the error channel instead, which could happen before the Dial returns. We only need to close the channel to un-block in a non-error scenario; otherwise it can be GCd without closing as we return on an error on the channel.